File: Base/Queue/Javascript/SeqQueue.js
class MyQueue {
  constructor() {
    this.MAXSIZE = 100;
    this.head = 0;
    this.tail = 0;
    this.queue = [];
  }

  // 往队列里面插入数据
  enqueue(item) {
    if (this.tail === this.MAXSIZE) {
      return;
    }
    this.queue[this.tail++] = item;
  }

  // pop 一个数据出队列
  dequeue() {
    if (this.head === this.tail) {
      return -1;
    }
    return this.queue[this.head++];
  }

  print() {
    let temp = [];
    for (let i = this.head; i < this.tail; i++) {
      temp.push(this.queue[i]);
    }
    console.log('all queue values: ', temp);
  }
}

let my_queue = new MyQueue();
my_queue.enqueue(1);
my_queue.enqueue(2);
my_queue.enqueue(3);
console.log(my_queue);
console.log('dequeue', my_queue.dequeue());
my_queue.print();

File: Queue/Rust/src/seq_queue.rs
use std::fmt::Display;

#[derive(Debug)]
pub struct SeqQueue<T> {
    maxsize: usize,
    head: usize,
    tail: usize,
    queue: Vec<T>,
}

impl<T> SeqQueue<T> {
    pub fn new() -> SeqQueue<T> {
        SeqQueue {
            maxsize: 100,
            head: 0,
            tail: 0,
            queue: Vec::with_capacity(100),
        }
    }

    pub fn enqueue(&mut self, item: T) -> bool {
        if self.tail == self.maxsize {
            return false;
        }
        self.queue.push(item);
        self.tail = self.tail + 1;

        true
    }

    pub fn dequeue(&mut self) -> Option<&T> {
        if self.head == self.tail {
            return None;
        }
        let data = &self.queue[self.head];
        self.head = self.head + 1;
        Option::Some(data)
    }
}

impl<T: Display> SeqQueue<T> {
    pub fn print(&self) {
        for i in self.head..self.tail {
            let value = &self.queue[i];
            println!("{}", value);
        }
    }
}